portfolio/src/main/webapp/charts.js

```diff
@@ -0,0 +1,109 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+google.charts.load('current', {'packages':['corechart']});
+google.charts.setOnLoadCallback(drawPoll);
+
+// Map that holds all polls that can be voted on.
+const polls = new Map([
+  [
+    'catsDogs', {
+      title: 'Do you Prefer Cats or Dogs?',
+      choices: ['Cats', 'Dogs'],
+    }
+  ],
+  [
+    'iceCream', {
+      title: 'What is the Best Ice Cream Flavor?',
+      choices: [
+        'Vanilla', 
+        'Chocolate', 
+        'Cookies and Cream', 
+        'Coffee', 
+        'Chocolate Chip',
+      ],
+    }
+  ],
+  [
+    'pancakesWaffles', {
+      title: 'Do you Prefer Pancakes or Waffles?',
+      choices: ['Pancakes', 'Waffles'],
+    }
+  ],
+])
+
+/**
+ * Generate html for the options of polls to choose from.
+ */
+function getPollOptions() {
+  return Array.from(polls).map((poll) => { return `
+    <option value="${poll[0]}">${poll[1].title}</option>
+  `
+  }).join('');
+}
+document.getElementById('polls-choice').innerHTML = getPollOptions();
+
+/**
+ * Create the input form for user voting.
+ */
+function createInputForm(poll) {
+  formOptions = polls.get(poll).choices.map(choice => { return `
+    <option value="${choice}">${choice}</option>
+  `
+  }).join('');
+
+  form = `
+    <form action="/vote-results?poll=${poll}" method="POST">
+      <select name="vote">
+        ${formOptions}
+      </select>
+      <button>Submit</button>
+    </form>
+  `;
+
+  document.getElementById('form-container').innerHTML = form;
+}
+
+/**
+ * Fetche the voting data and creates a chart with it,
+ * also update the voting submission form.
+ */
+async function drawPoll() {
+  // Determine which poll we want to display.
+  const selectorElement = document.getElementById('polls-choice');
+  const choice = selectorElement.options[selectorElement.selectedIndex].value;
+
+  const response = await fetch(`/vote-results?poll=${choice}`);
```

Comment:
It's always good to URL-encode strings before putting them in a URL using the built-in encodeURIComponent function. I don't think this is a security problem but it can certainly cause bugs (for example if you accidentally name a poll with a `?` or `&` character in it. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent

Reply:
Thank! Great feedback, fix is pushed.